File: lib/core/global/interfaces/rest_client.dart
import 'package:sertton/core/global/responses/rest_response.dart';
import 'package:sertton/rest/types/json.dart';

abstract class RestClient {
  Future<RestResponse<Json>> get(String path, {Json? queryParams});
  Future<RestResponse<Json>> post(String path, {Json? body, Json? queryParams});
  Future<RestResponse<Json>> put(String path, {Json? body, Json? queryParams});
  Future<RestResponse<Json>> delete(
    String path, {
    Json? body,
    Json? queryParams,
  });
  void getBaseUrl();
  void setBaseUrl(String baseUrl);
  void setHeader(String key, String value);
}

File: src/core/shared/interfaces/rest-client.ts
import type { RestResponse } from "@/core/shared/responses/rest-response"
import type { JsonObject } from "@/core/shared/types/json"

export interface RestRequestOptions {
  readonly body?: JsonObject
  readonly headers?: Readonly<Record<string, string>>
  readonly query?: Readonly<Record<string, string | number | boolean | undefined>>
  readonly signal?: AbortSignal
}

export interface RestClient {
  clearQueryParams(): void
  delete<Response>(path: string, options?: RestRequestOptions): Promise<RestResponse<Response>>
  get<Response>(path: string, options?: RestRequestOptions): Promise<RestResponse<Response>>
  post<Response>(path: string, options?: RestRequestOptions): Promise<RestResponse<Response>>
  put<Response>(path: string, options?: RestRequestOptions): Promise<RestResponse<Response>>
  setQueryParam(key: string, value: string | number | boolean | undefined): void
}

export type IRestClient = RestClient
